File: app/src/main/java/com/benny/openlauncher/viewutil/DialogHelper.java
package com.benny.openlauncher.viewutil;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v7.widget.LinearLayoutManager;
import android.view.Gravity;
import android.view.View;

import com.afollestad.materialdialogs.MaterialDialog;
import com.benny.openlauncher.R;
import com.benny.openlauncher.activity.HomeActivity;
import com.benny.openlauncher.model.App;
import com.benny.openlauncher.model.Item;
import com.benny.openlauncher.util.AppManager;
import com.benny.openlauncher.util.AppSettings;
import com.benny.openlauncher.util.Tool;
import com.mikepenz.fastadapter.IAdapter;
import com.mikepenz.fastadapter.commons.adapters.FastItemAdapter;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class DialogHelper {
    public static void editItemDialog(String title, String defaultText, Context c, final OnItemEditListener listener) {
        MaterialDialog.Builder builder = new MaterialDialog.Builder(c);
        builder.title(title)
                .positiveText(android.R.string.ok)
                .negativeText(android.R.string.cancel)
                .input(null, defaultText, new MaterialDialog.InputCallback() {
                    @Override
                    public void onInput(@NonNull MaterialDialog dialog, CharSequence input) {
                        listener.itemLabel(input.toString());
                    }
                }).show();
    }

    public static void alertDialog(Context context, String title, String msg, MaterialDialog.SingleButtonCallback onPositive) {
        MaterialDialog.Builder builder = new MaterialDialog.Builder(context);
        builder.title(title)
                .onPositive(onPositive)
                .content(msg)
                .negativeText(android.R.string.cancel)
                .positiveText(android.R.string.ok)
                .show();
    }

    public static void alertDialog(Context context, String title, String message, String positive, MaterialDialog.SingleButtonCallback onPositive) {
        MaterialDialog.Builder builder = new MaterialDialog.Builder(context);
        builder.title(title)
                .onPositive(onPositive)
                .content(message)
                .negativeText(android.R.string.cancel)
                .positiveText(positive)
                .show();
    }

    public static void selectActionDialog(final Context context, MaterialDialog.ListCallback callback) {
        MaterialDialog.Builder builder = new MaterialDialog.Builder(context);
        builder.title(R.string.action)
                .items(R.array.entries__gesture_action)
                .itemsCallback(callback)
                .show();
    }

    public static void selectDesktopActionDialog(final Context context, MaterialDialog.ListCallback callback) {
        MaterialDialog.Builder builder = new MaterialDialog.Builder(context);
        builder.title(R.string.action)
                .items(R.array.entries__desktop_actions)
                .itemsCallback(callback)
                .show();
    }

    public static void selectGestureDialog(final Context context, String title, MaterialDialog.ListCallback callback) {
        MaterialDialog.Builder builder = new MaterialDialog.Builder(context);
        builder.title(title)
                .items(R.array.entries__gesture)
                .itemsCallback(callback)
                .show();
    }

    public static void selectAppDialog(final Context context, final OnAppSelectedListener onAppSelectedListener) {
        MaterialDialog.Builder builder = new MaterialDialog.Builder(context);
        FastItemAdapter<IconLabelItem> fastItemAdapter = new FastItemAdapter<>();
        builder.title(R.string.select_app)
                .adapter(fastItemAdapter, new LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false))
                .negativeText(android.R.string.cancel);

        final MaterialDialog dialog = builder.build();
        List<IconLabelItem> items = new ArrayList<>();
        final List<App> apps = AppManager.getInstance(context).getApps();
        for (int i = 0; i < apps.size(); i++) {
            items.add(new IconLabelItem(apps.get(i).getIcon(), apps.get(i).getLabel())
                    .withIconSize(50)
                    .withIsAppLauncher(true)
                    .withIconGravity(Gravity.START)
                    .withIconPadding(8));
        }
        fastItemAdapter.set(items);
        fastItemAdapter.withOnClickListener(new com.mikepenz.fastadapter.listeners.OnClickListener<IconLabelItem>() {
            @Override
            public boolean onClick(View v, IAdapter<IconLabelItem> adapter, IconLabelItem item, int position) {
                if (onAppSelectedListener != null) {
                    onAppSelectedListener.onAppSelected(apps.get(position));
                }
                dialog.dismiss();
                return true;
            }
        });
        dialog.show();
    }

    public static void startPickIconPackIntent(final Context context) {
        PackageManager packageManager = context.getPackageManager();
        Activity activity = (Activity) context;
        AppManager appManager = AppManager.getInstance(context);
        Intent intent = new Intent(Intent.ACTION_MAIN);
        intent.addCategory("com.anddoes.launcher.THEME");

        FastItemAdapter<IconLabelItem> fastItemAdapter = new FastItemAdapter<>();

        final List<ResolveInfo> resolveInfos = packageManager.queryIntentActivities(intent, 0);
        Collections.sort(resolveInfos, new ResolveInfo.DisplayNameComparator(packageManager));
        final MaterialDialog dialog = new MaterialDialog.Builder(activity)
                .adapter(fastItemAdapter, null)
                .title((activity.getString(R.string.select_icon_pack)))
                .build();

        fastItemAdapter.add(new IconLabelItem(activity, R.mipmap.ic_launcher, R.string.default_icons)
                .withIconPadding(16)
                .withIconGravity(Gravity.START)
                .withOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        appManager._recreateAfterGettingApps = true;
                        AppSettings.get().setIconPack("");
                        appManager.getAllApps();
                        dialog.dismiss();
                    }
                }));
        for (int i = 0; i < resolveInfos.size(); i++) {
            final int mI = i;
            fastItemAdapter.add(new IconLabelItem(resolveInfos.get(i).loadIcon(packageManager), resolveInfos.get(i).loadLabel(packageManager).toString())
                    .withIconPadding(16)
                    .withIconSize(50)
                    .withIsAppLauncher(true)
                    .withIconGravity(Gravity.START)
                    .withOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
                                appManager._recreateAfterGettingApps = true;
                                AppSettings.get().setIconPack(resolveInfos.get(mI).activityInfo.packageName);
                                appManager.getAllApps();
                                dialog.dismiss();
                            } else {
                                Tool.toast(context, (activity.getString(R.string.toast_icon_pack_error)));
                                ActivityCompat.requestPermissions(HomeActivity.Companion.getLauncher(), new String[]{Manifest.permission.READ_EXTERNAL_STORAGE}, HomeActivity.REQUEST_PERMISSION_STORAGE);
                            }
                        }
                    }));
        }
        dialog.show();
    }

    public static void deletePackageDialog(Context context, Item item) {
        if (item.getType() == Item.Type.APP) {
            try {
                Uri packageURI = Uri.parse("package:" + item.getIntent().getComponent().getPackageName());
                Intent uninstallIntent = new Intent(Intent.ACTION_DELETE, packageURI);
                context.startActivity(uninstallIntent);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public interface OnAppSelectedListener {
        void onAppSelected(App app);
    }

    public interface OnItemEditListener {
        void itemLabel(String label);
    }
}

File: app/src/main/java/com/benny/openlauncher/viewutil/DialogHelper.kt
package com.benny.openlauncher.viewutil

import android.app.WallpaperManager
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Environment
import android.support.v7.widget.LinearLayoutManager
import android.view.DragEvent
import android.view.Gravity
import android.widget.Toast

import com.afollestad.materialdialogs.MaterialDialog
import com.benny.openlauncher.R
import com.benny.openlauncher.core.model.IconLabelItem
import com.benny.openlauncher.core.model.Item
import com.benny.openlauncher.core.util.DragNDropHandler
import com.benny.openlauncher.core.util.Tool
import com.benny.openlauncher.util.AppManager
import com.benny.openlauncher.util.LauncherAction
import com.benny.openlauncher.util.copy
import com.benny.openlauncher.util.getStartAppIntent
import com.mikepenz.fastadapter.commons.adapters.FastItemAdapter

import net.qiujuer.genius.blur.StackBlur

import java.io.File
import java.util.ArrayList

object DialogHelper {
    fun editItemDialog(title: String, defaultText: String, c: Context, listener: OnItemEditListener) {
        val builder = MaterialDialog.Builder(c)
        builder.title(title)
                .positiveText(R.string.ok)
                .negativeText(R.string.cancel)
                .input(null, defaultText) { dialog, input -> listener.itemLabel(input.toString()) }.show()
    }

    fun alertDialog(context: Context, title: String, msg: String) {
        val builder = MaterialDialog.Builder(context)
        builder.title(title)
                .content(msg)
                .negativeText(R.string.cancel)
                .positiveText(R.string.ok)
                .show()
    }

    fun alertDialog(context: Context, title: String, msg: String, onPositive: MaterialDialog.SingleButtonCallback) {
        val builder = MaterialDialog.Builder(context)
        builder.title(title)
                .onPositive(onPositive)
                .content(msg)
                .negativeText(R.string.cancel)
                .positiveText(R.string.ok)
                .show()
    }

    fun alertDialog(context: Context, title: String, msg: String, positive: String, onPositive: MaterialDialog.SingleButtonCallback) {
        val builder = MaterialDialog.Builder(context)
        builder.title(title)
                .onPositive(onPositive)
                .content(msg)
                .negativeText(R.string.cancel)
                .positiveText(positive)
                .show()
    }

    fun addActionItemDialog(context: Context, callback: MaterialDialog.ListCallback) {
        val builder = MaterialDialog.Builder(context)
        builder.title(R.string.desktop_action)
                .items(R.array.entries__desktop_actions)
                .itemsCallback(callback)
                .show()
    }

    fun selectAppDialog(context: Context, onAppSelectedListener: OnAppSelectedListener?) {
        val builder = MaterialDialog.Builder(context)
        val fastItemAdapter = FastItemAdapter<IconLabelItem>()
        builder.title(R.string.select_app)
                .adapter(fastItemAdapter, LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false))
                .negativeText(R.string.cancel)

        val dialog = builder.build()
        val items = ArrayList<IconLabelItem>()
        val apps = AppManager.getInstance(context).apps
        val size = Tool.dp2px(18, context)
        val sizePad = Tool.dp2px(8, context)
        for (i in apps.indices) {
            items.add(IconLabelItem(context, apps[i].iconProvider, apps[i].label, size)
                    .withIconGravity(Gravity.START)
                    .withDrawablePadding(context, sizePad))
        }
        fastItemAdapter.set(items)
        fastItemAdapter.withOnClickListener { v, adapter, item, position ->
            onAppSelectedListener?.onAppSelected(apps[position])
            dialog.dismiss()
            true
        }
        dialog.show()
    }

    fun selectActionDialog(context: Context, title: String, selected: LauncherAction.ActionItem, onActionSelectedListener: OnActionSelectedListener?) {
        MaterialDialog.Builder(context)
                .title(title)
                .negativeText(R.string.cancel)
                .items(R.array.entries__gestures)
                .itemsCallbackSingleChoice(LauncherAction.getActionItemIndex(selected) + 1) { dialog, itemView, which, text ->
                    var item: LauncherAction.ActionItem? = null
                    if (which > 0) {
                        item = LauncherAction.getActionItem(which - 1)
                        if (item != null && item.action == LauncherAction.Action.LaunchApp) {
                            val finalItem = item
                            selectAppDialog(context, object : OnAppSelectedListener {
                                override fun onAppSelected(app: AppManager.App) {
                                    finalItem.extraData = Tool.getStartAppIntent(app)
                                    onActionSelectedListener!!.onActionSelected(finalItem)
                                }
                            })
                        } else onActionSelectedListener?.onActionSelected(item)
                    }
                    true
                }.show()
    }

    fun setWallpaperDialog(context: Context) {
        MaterialDialog.Builder(context)
                .title(R.string.wallpaper)
                .iconRes(R.drawable.ic_photo_black_24dp)
                .items(R.array.entries__wallpaper_options)
                .itemsCallback { dialog, itemView, position, text ->
                    when (position) {
                        0 -> {
                            val intent = Intent(Intent.ACTION_SET_WALLPAPER)
                            context.startActivity(Intent.createChooser(intent, context.getString(R.string.wallpaper_pick)))
                        }
                        1 -> try {
                            WallpaperManager.getInstance(context).setBitmap(StackBlur.blur(Tool.drawableToBitmap(context.wallpaper), 10, false))
                        } catch (e: Exception) {
                            Tool.toast(context, context.getString(R.string.wallpaper_unable_to_blur))
                        }

                    }
                }.show()
    }

    fun deletePackageDialog(context: Context, item: Item) {
        if (item.type == Item.Type.APP) {
            try {
                val packageURI = Uri.parse("package:" + item.intent?.component!!.packageName)
                val uninstallIntent = Intent(Intent.ACTION_DELETE, packageURI)
                context.startActivity(uninstallIntent)
            } catch (e: Exception) {

            }
        }
    }

    fun backupDialog(context: Context) {
        MaterialDialog.Builder(context)
                .title(R.string.pref_title__backup)
                .positiveText(R.string.cancel)
                .items(R.array.entries__backup_options)
                .itemsCallback { dialog, itemView, item, text ->
                    val m = context.packageManager
                    var s = context.packageName

                    if (context.resources.getStringArray(R.array.entries__backup_options)[item] == context.resources.getString(R.string.dialog__backup_app_settings__backup)) {
                        val directory = File(Environment.getExternalStorageDirectory().toString() + "/OpenLauncher/")
                        if (!directory.exists()) {
                            directory.mkdir()
                        }
                        try {
                            val p = m.getPackageInfo(s, 0)
                            s = p.applicationInfo.dataDir
                            Tool.copy(context, s + "/databases/home.db", directory.toString() + "/home.db")
                            Tool.copy(context, s + "/shared_prefs/app.xml", directory.toString() + "/app.xml")
                            Toast.makeText(context, R.string.dialog__backup_app_settings__success, Toast.LENGTH_SHORT).show()
                        } catch (e: Exception) {
                            Toast.makeText(context, R.string.dialog__backup_app_settings__error, Toast.LENGTH_SHORT).show()
                        }

                    }
                    if (context.resources.getStringArray(R.array.entries__backup_options)[item] == context.resources.getString(R.string.dialog__backup_app_settings__restore)) {
                        val directory = File(Environment.getExternalStorageDirectory().toString() + "/OpenLauncher/")
                        try {
                            val p = m.getPackageInfo(s, 0)
                            s = p.applicationInfo.dataDir
                            Tool.copy(context, directory.toString() + "/home.db", s + "/databases/home.db")
                            Tool.copy(context, directory.toString() + "/app.xml", s + "/shared_prefs/app.xml")
                            Toast.makeText(context, R.string.dialog__backup_app_settings__success, Toast.LENGTH_SHORT).show()
                        } catch (e: Exception) {
                            Toast.makeText(context, R.string.dialog__backup_app_settings__error, Toast.LENGTH_SHORT).show()
                        }

                        System.exit(1)
                    }
                }.show()
    }

    interface OnAppSelectedListener {
        fun onAppSelected(app: AppManager.App)
    }

    interface OnActionSelectedListener {
        fun onActionSelected(item: LauncherAction.ActionItem?)
    }

    interface OnItemEditListener {
        fun itemLabel(label: String)
    }
}
